pilot/pkg/networking/core/v1alpha3/gateway_test.go

```diff
@@ -2284,3 +2286,194 @@ func TestBuildNameToServiceMapForHttpRoutes(t *testing.T) {
+						Gateways: []string{"testns/terminated-tls-gateway"},
+						Hosts:    []string{"barone.example.com"},
+						Tcp: []*networking.TCPRoute{
+							{
+								Match: []*networking.L4MatchAttributes{
+									{
+										Port: 5678,
+									},
+								},
+								Route: []*networking.RouteDestination{
+									{
+										Destination: &networking.Destination{
+											Host: "foo.com",
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			expectedHTTPFilters:    []string{},
+			expectedNetworkFilters: []string{wellknown.TCPProxy},
+		},
+		{
+			name: "non-http istio-mtls server",
+			gateways: []config.Config{
+				{
+					Meta: config.Meta{Name: "non-http-gateway", Namespace: "testns", GroupVersionKind: gvk.Gateway},
+					Spec: &networking.Gateway{
+						Servers: []*networking.Server{
+							{
+								Port:  &networking.Port{Name: "mtls", Number: 15443, Protocol: "TLS"},
+								Hosts: []string{"barone.example.com"},
+								Tls:   &networking.ServerTLSSettings{Mode: networking.ServerTLSSettings_ISTIO_MUTUAL},
+							},
+						},
+					},
+				},
+			},
+			virtualServices: []config.Config{
+				{
+					Meta: config.Meta{Name: uuid.NewString(), Namespace: uuid.NewString(), GroupVersionKind: gvk.VirtualService},
+					Spec: &networking.VirtualService{
+						Gateways: []string{"testns/non-http-gateway"},
+						Hosts:    []string{"barone.example.com"},
+						Tcp: []*networking.TCPRoute{
+							{
+								Match: []*networking.L4MatchAttributes{
+									{
+										Port: 15443,
+									},
+								},
+								Route: []*networking.RouteDestination{
+									{
+										Destination: &networking.Destination{
+											Host: "foo.com",
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			expectedHTTPFilters:    []string{},
+			expectedNetworkFilters: []string{xdsfilters.TCPListenerMx.GetName(), wellknown.TCPProxy},
+		},
+	}
+	for _, tt := range cases {
+		t.Run(tt.name, func(t *testing.T) {
+			Configs := make([]config.Config, 0)
+			Configs = append(Configs, tt.gateways...)
+			Configs = append(Configs, tt.virtualServices...)
+			cg := NewConfigGenTest(t, TestOptions{
+				Configs: Configs,
+			})
+			proxy := cg.SetupProxy(&proxyGateway)
+			proxy.Metadata = &proxyGatewayMetadata
+
+			builder := cg.ConfigGen.buildGatewayListeners(&ListenerBuilder{node: proxy, push: cg.PushContext()})
+			for _, listener := range builder.gatewayListeners {
+				for _, fc := range listener.GetFilterChains() {
+					gotNWFilters, gotHTTPFilters := xdstest.ExtractFilterNames(t, fc)
+					if !reflect.DeepEqual(gotHTTPFilters, tt.expectedHTTPFilters) {
+						t.Fatalf("Expected HTTPfilters: %v, got: %v", tt.expectedHTTPFilters, gotHTTPFilters)
+					}
+					if !reflect.DeepEqual(gotNWFilters, tt.expectedNetworkFilters) {
+						t.Fatalf("Expected network filters: %v, got: %v", tt.expectedNetworkFilters, gotNWFilters)
+					}
+				}
+			}
+		})
+	}
```

Comment:
```suggestion
			listenertest.VerifyListeners(t, builder.gatewayListeners, listenertest.ListenersTest{
				Listener: listenertest.ListenerTest{FilterChains: []listenertest.FilterChainTest{
					{
						NetworkFilters: tt.expectedNetworkFilters,
						HTTPFilters:    tt.expectedHTTPFilters,
					},
				}}})
```

nit: could use the new matching framework for listeners. Doesn't save much code here but its nice to get more usage. Also the current test would pass if no listeners were returned.